lib/premailer/adapter/nokogiri.rb

```diff
@@ -150,9 +150,9 @@ def write_unmergable_css_rules(doc, unmergable_rules) # :nodoc:
           style_tag = "<style type=\"text/css\">\n#{styles}</style>"
           unless (body = doc.search('body')).empty?
             if doc.at_css('body').children && !doc.at_css('body').children.empty?
-              doc.at_css('body').children.before(::Nokogiri::XML.fragment(style_tag))
+              doc.at_css('head').children.after(::Nokogiri::XML.fragment(style_tag))
```

Comment:
These conditions are conflicting: "if there's a body, put it in the head".

If anything, this should be checking for presence of `<head>` and add it if needed.